### PR TITLE
fix: various fixes

### PR DESCRIPTION
**Pull request by falconandy**
### Summary

Various fixes, including support for:
* [x] minIdLength
* [x] serverZone
* [x] batch API (currently without count/size limits)
* [x] callback support  

Available as version `v0.0.3`

### Checklist

* [x] Does your PR title have the correct [title format](https://github.com/amplitude/analytics-go/blob/main/CONTRIBUTING.md#pr-commit-title-conventions)?
* Does your PR have a breaking change?:  Yes

### CODE REVIEW

File: amplitude/amplitude_client.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"net/http"
+	"time"
+)
+
+func newAmplitudeClient(
```
Comment (liuyang1520) [inline amplitude/amplitude_client.go:11]: Probably I need some background on this, I think `amplitudeClient` is a wrapped HTTP client, which is different from the SDK client in `client.go`. If yes, should we make it more clear with `amplitude_http_client` name?

File: amplitude/amplitude_client.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	serverURL string, options clientPayloadOptions, logger Logger, connectionTimeout time.Duration,
+) *amplitudeClient {
+	var payloadOptions *clientPayloadOptions
+	if options != (clientPayloadOptions{}) {
```
Comment (liuyang1520) [inline amplitude/amplitude_client.go:15]: Minor, I might be wrong, maybe just use `options` directly in `amplitudeClient`? I am not sure what is the difference.

File: amplitude/amplitude_client.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+		}
+	}()
+
+	h.logger.Infof("HTTP response code: %s", response.Status)
+	body, err := io.ReadAll(response.Body)
+	if err != nil {
+		h.logger.Warnf("HTTP response, can't read body: %s", err)
+	} else {
+		h.logger.Infof("HTTP response body: %s", string(body))
+	}
```
Comment (liuyang1520) [inline amplitude/amplitude_client.go:90]: Like the error handlings here, good job!

File: amplitude/client.go [hunk 4]
```diff
@@ -1,61 +1,108 @@
 package amplitude
 
+import (
+	"github.com/amplitude/analytics-go/amplitude/internal"
```
Comment (liuyang1520) [inline amplitude/client.go:4]: Not sure why we are importing from github link, should we import with relative local path here?

File: amplitude/context_plugin.go [hunk 5]
```diff
@@ -36,7 +40,9 @@ func (c *ContextPlugin) Execute(event *Event) *Event {
 		event.InsertID = uuid.NewString()
 	}
 
-	event.Library = c.contextString
+	event.Library = p.contextString
 
 	return event
 }
+
+var _ EnrichmentPlugin = (*ContextPlugin)(nil)
```
Comment (liuyang1520) [inline amplitude/context_plugin.go:48]: Curious what is this line doing?

File: amplitude/amplitude_client.go (hunk 1 above)
Comment (falconandy) [reply amplitude/amplitude_client.go]: Renamed

File: amplitude/amplitude_client.go (hunk 2 above)
Comment (falconandy) [reply amplitude/amplitude_client.go]: To not serialize empty options `options: {}` and save some payload size.

File: amplitude/context_plugin.go (hunk 5 above)
Comment (falconandy) [reply amplitude/context_plugin.go]: Compile-time check that `ContextPlugin` implements `EnrichmentPlugin` interface. Unnecessary now (in `v0.0.4`) - removed.

Comment (liuyang1520) [reply amplitude/context_plugin.go]: Got it, thanks for the explanation!

**Review by liuyang1520 — APPROVED** [review]
Besides the github import mentioned in previous comment, others LGTM

File: amplitude/client.go (hunk 4 above)
Comment (falconandy) [reply amplitude/client.go:5]: Go doesn't support relative paths

**Review by justin-fiedler — CHANGES_REQUESTED** [review]

**Comment by falconandy** [comment]
@justin-fiedler In Ampli codegen Variadic Functions are used to emulate optional parameter:
```
func (a *Ampli) Track(userID string, event Event, eventOptions ...EventOptions) {
	var options EventOptions
	if len(eventOptions) > 0 {
		options = eventOptions[0]
	}
	...
```

Usage:
```
ampli.Track(userId, event)
ampli.Track(userId, event, options)
```

File: amplitude/plugins/enrichment/context_plugin.go [hunk 6]
```diff
@@ -12,22 +15,26 @@ type ContextPlugin struct {
 	contextString string
 }
 
-func NewContextPlugin() *ContextPlugin {
+func NewContextPlugin() types.EnrichmentPlugin {
```
[the remark this answers is not in the document]
Comment (falconandy) [reply amplitude/plugins/enrichment/context_plugin.go]: Fixed

[the remark this answers is not in the document]
Comment (justin-fiedler) [reply amplitude/plugins/enrichment/context_plugin.go]: Before and Enrichment are two different types. See link above

[the remark this answers is not in the document]
Comment (falconandy) [reply amplitude/plugins/enrichment/context_plugin.go]: Fixed

**Review by justin-fiedler — APPROVED** [review]

**Pull request merged**